h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostModel.java

```diff
@@ -1,32 +1,163 @@
+    assert treeNodes.length >= 1;
+
+    SharedTreeGraph sharedTreeGraph = new SharedTreeGraph();
+    final SharedTreeSubgraph sharedTreeSubgraph = sharedTreeGraph.makeSubgraph(_output._training_metrics._description);
+    treeNodes[0].split_index();
+
+    final FeatureProperties featureProperties = assembleFeatureNames(model_info._dataInfoKey.get()); // XGBoost's usage of one-hot encoding assumed
+    constructSubgraph(treeNodes, sharedTreeSubgraph.makeRootNode(), 0, sharedTreeSubgraph, featureProperties, true); // Root node is at index 0
+    return sharedTreeGraph;
+  }
+
+  private static void constructSubgraph(final RegTreeNode[] xgBoostNodes, final SharedTreeNode sharedTreeNode,
+                                        final int nodeIndex, final SharedTreeSubgraph sharedTreeSubgraph,
+                                        final FeatureProperties featureProperties, boolean inclusiveNA) {
+    final RegTreeNode xgBoostNode = xgBoostNodes[nodeIndex];
+    // Not testing for NaNs, as SharedTreeNode uses NaNs as default values.
+    //No domain set, as the structure mimics XGBoost's tree, which is numeric-only
+    if (featureProperties._oneHotEncoded[xgBoostNode.split_index()]) {
+      //Shared tree model uses < to the left and >= to the right. Transforiming one-hot encoded categoricals
+      // from 0 to 1 makes it fit the current split description logic
+      sharedTreeNode.setSplitValue(1.0F);
+    } else {
+      sharedTreeNode.setSplitValue(xgBoostNode.getSplitCondition());
+    }
+    sharedTreeNode.setPredValue(xgBoostNode.getLeafValue());
+    sharedTreeNode.setCol(xgBoostNode.split_index(), featureProperties._names[xgBoostNode.split_index()]);
+    sharedTreeNode.setInclusiveNa(inclusiveNA);
+    sharedTreeNode.setNodeNumber(nodeIndex);
+
+    if (xgBoostNode.getLeftChildIndex() != -1) {
+      constructSubgraph(xgBoostNodes, sharedTreeSubgraph.makeLeftChildNode(sharedTreeNode),
+              xgBoostNode.getLeftChildIndex(), sharedTreeSubgraph, featureProperties, xgBoostNode.default_left());
+    }
+
+    if (xgBoostNode.getRightChildIndex() != -1) {
+      constructSubgraph(xgBoostNodes, sharedTreeSubgraph.makeRightChildNode(sharedTreeNode),
+              xgBoostNode.getRightChildIndex(), sharedTreeSubgraph, featureProperties, !xgBoostNode.default_left());
+    }
+  }
+
+
+  private static FeatureProperties assembleFeatureNames(final DataInfo di) {
+    String[] coefnames = di.coefNames();
+    assert (coefnames.length == di.fullN());
+    int numCatCols = di._catOffsets[di._catOffsets.length - 1];
+
+    String[] featureNames = new String[di.fullN()];
+    boolean[] oneHotEncoded = new boolean[di.fullN()];
+    for (int i = 0; i < di.fullN(); ++i) {
+      featureNames[i] = coefnames[i];
+      if (i < numCatCols) {
+        oneHotEncoded[i] = true;
+      }
+    }
+
+    return new FeatureProperties(featureNames, oneHotEncoded);
+  }
+
+  private static class FeatureProperties {
```

Comment:
it seems that a HashMap or LinkedHashMap would replace this structure

Reply:
Not really. There is a need to return two arrays of different types. The arrays are accessed by index, which is not calculated anywhere, as this is naturally provided by XGBoost's tree representation. 

They key points to these two arrrays, one of them indicated if the feature is result of one-hot encoding and the second one is name of the feature. This points to a MultiMap solution. However, a multimap basically maps one key to multiple values of the **same type**. Which is not the case here and additional type checks and conversion would be required. It could be done using the same type, but unnecessary conversions would be required. The current access is reasonably fast (faster than using a map) and suits the purpose very well.

It could be done, but I don't see any advantage.